lib/spike.rb

```diff
@@ -0,0 +1,361 @@
+    end
+
+    def api_path
+      path = definition.path
+      path = path.gsub("/repos/:owner/:repo", "\#{Repository.path repo}")
+      path = definition.params.select(&:required).reduce(path) do |path, param|
+        path.gsub(":#{param.name}", "\#{#{param.name}}")
+      end
+    end
+
+    def return_type_description
+      if definition.verb == "GET" && !resource.singular?
+        "[Array<Sawyer::Resource>]"
+      else
+        "<Sawyer::Resource>"
+      end
+    end
+
+    def parameter_type(parameter)
+      {
+        "repo" => "[Integer, String, Repository, Hash]",
+      }[parameter.name] || "[#{parameter.type.capitalize}]"
+    end
+
+    def parameter_description(parameter)
+      return parameter.description unless parameter.description.empty?
+      return "A GitHub repository" if parameter.name == "repo"
+
+      "The ID of the #{parameter.name.gsub("_id", "").gsub("_", " ")}" if parameter.name.end_with?("_id")
+    end
+
+    def parameter_documentation
+      definition.params.select(&:required).map {|param|
+        "@param #{param.name} #{parameter_type(param)} #{parameter_description(param)}"
+      } + definition.params.reject(&:required).reject {|param| ["per_page", "page"].include?(param.name)}.map {|param|
+        "@param options [#{param.type.capitalize}] :#{param.name} #{param.description}"
+      }
+    end
+
+    def return_value_description
+      case verb
+      when "GET"
+        if resource.singular?
+          "A single #{resource.name.gsub("_", " ")}"
+        else
+          "A list of #{resource.name.gsub("_", " ")}"
+        end
+      when "POST"
+        "The new #{resource.name.singularize.gsub("_", " ")}"
+      else
+      end
+    end
+
+    def verb
+      definition["method"]
+    end
+
+    def parameters
+      parameterizer.parameterize(arguments)
+    end
+
+    def arguments
+      definition.params.select(&:required).map(&:name)
+    end
+
+    def resource
+      return @resource if defined?(@resource)
+
+      final_segment = path_segments.reject {|segment| segment.start_with?(":")}.last
+      if final_segment == directory
+        @resource = final_segment
+      else
+        @resource = [directory.singularize, final_segment].join("_")
+      end
+    end
+
+    def path_segments
+      @path_segments ||= definition.path.to_s.split("/").reject(&:empty?)
+    end
+
+    def method_name
+      case verb
+      when "GET"
+        resource.name
+      when "POST"
+        "create_#{resource.name.singularize}"
+      else
+      end
+    end
+
+    def alternate_name
+      return unless definition.verb == "GET"
+      return if resource.singular?
+      "list_#{resource.name}"
+    end
+
+    def parts
+      @parts ||= path_segments[path_segments.index(directory)..-1].reject {|segment| segment.start_with?(":")}
+    end
+
+    def priority
+      [parts.count, VERB_PRIORITY.index(verb), resource.singular?? 0 : 1]
+    end
```

Comment:
This is so we can say `sort_by(&:priority)` on the endpoints to get the method definitions in the right order.